src/napi/node_api_async.c

```diff
@@ -0,0 +1,169 @@
+/* Copyright 2019-present Samsung Electronics Co., Ltd. and other contributors
+ * Copyright 2018-present Rokid Co., Ltd. and other contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "uv.h"
+#include "internal/node_api_internal.h"
+
+static void iotjs_uv_work_cb(uv_work_t* req) {
+  iotjs_async_work_t* async_work = (iotjs_async_work_t*)req->data;
+  IOTJS_ASSERT(async_work != NULL);
+  if (async_work->execute != NULL) {
+    async_work->execute(async_work->env, async_work->data);
+  }
+}
+
+static void iotjs_uv_work_after_cb(uv_work_t* req, int status) {
+  iotjs_async_work_t* async_work = (iotjs_async_work_t*)req->data;
+  IOTJS_ASSERT(async_work != NULL);
+  napi_status cb_status;
+  if (status == 0) {
+    cb_status = napi_ok;
+  } else if (status == UV_ECANCELED) {
+    cb_status = napi_cancelled;
+  } else {
+    cb_status = napi_generic_failure;
+  }
+
+  napi_env env = async_work->env;
+  napi_async_complete_callback complete = async_work->complete;
+  void* data = async_work->data;
+
+  if (complete != NULL) {
+    jerryx_handle_scope scope;
+    jerryx_open_handle_scope(&scope);
+    /**
+     * napi_async_work could be deleted by invocation of `napi_delete_asyncwork`
+     * in its complete callback.
+     */
+    complete(env, cb_status, data);
+    jerryx_close_handle_scope(scope);
+
+    if (iotjs_napi_is_exception_pending(env)) {
+      jerry_value_t jval_err;
+      jval_err = iotjs_napi_env_get_and_clear_exception(env);
+      if (jval_err == (uintptr_t)NULL) {
```

Comment:
This check is not correct. You cannot compare a `jerry_value_t` to a pointer. This leads to undefined behavoir, but with current JerryScript it is probably always false.

Reply:
`iotjs_napi_env_get_and_clear_exception` returns a `jerry_value_t`.
The function gets the current environments' error value with the `AS_JERRY_VALUE` macro. If it is `NULL`(`0x0`), the macro casts it to an `uintptr_t` then to a `jerry_value_t` which will hold the value of 0. Comparing `0` to `(uintptr_t)NULL` results in basically comparing `0` to `0`. If the `jerry_value_t` actually holds an error, the comparison returns false, since a jerry_value_t will guaranteed always have a value other than `0`.

Reply:
IMHO it is not a safe usage of the JerryScript API. `iotjs_napi_env_get_and_clear_exception` should be updated to use `jerry_create_undefined` or `jerry_create_null`.

Reply:
I can definitely do an update then.
Just a note:
There are other locations in the code (ie: node_api_function.c) using the same method.

Reply:
@DanielBallaSZTE we should fix them in a followup patch.

Reply:
Yes, it's unsafe. But there is no `Empty` value in JerryScript AFAIK, it's just a not so even bad choice.

Also note that we could throw any value in JS so as `undefined` and `null`.

Reply:
You convinced me. Leave as is for now.